### PR TITLE
build: Upgrade to React Native 0.73 (Android)

### PR DESCRIPTION
**Pull request by geriux**
Part of https://github.com/WordPress/gutenberg/pull/58475

## What?
This PR adds the required Android changes for the React Native `0.73.3` upgrade.

## Why?
To keep up to date with the latest stable releases of React Native.

## How?
- Migrates the demo app java files into Kotlin files, this was done using Android Studio helper.
- Includes Android 14 support which updates:
  - API Level to `34`
  - AGP to `8.1.1`
  - Kotlin to `1.8.0`
- Changes JDK to `17`
- Updates NDK to `25.1.8937393`
- Updates the React Native Libraries to `v5` 
- Updates the build gradle files following the [upgrade helper tool](https://react-native-community.github.io/upgrade-helper/?from=0.71.15&to=0.73.3)

## Testing Instructions
CI checks related to Android should succeed.

### CODE REVIEW

File: packages/react-native-editor/android/app/build.gradle
Comment (geriux) [inline packages/react-native-editor/android/app/build.gradle:108]: `v5` will be [released in this PR](https://github.com/wordpress-mobile/react-native-libraries-publisher/pull/33).

File: packages/react-native-editor/package.json
Comment (geriux) [inline packages/react-native-editor/package.json:37]: This is removed since React Native's has it as a dependency which supports the current AGP version.

File: packages/react-native-editor/react-native.config.js
Comment (geriux) [inline packages/react-native-editor/react-native.config.js]: These changes are needed because we finally added [this code](https://github.com/WordPress/gutenberg/pull/58551/files#diff-a13706be11b93908569684c9aed9cf0dbc05d11e254ea5dc01766344d203b1f2R133).

Where `applyNativeModulesAppBuildGradle` is called, this is needed to add extra configuration like references to [PackageList](https://github.com/WordPress/gutenberg/pull/58551/files#diff-2bba332a9cc79ca4865108adb419241e176da1bcd16bb8f1bb5e4312f43be4caR59). Since that code also handles auto-linking, we need to reference the libraries we are linking manually, currently all of them.

Comment (dcalhoun) [reply packages/react-native-editor/react-native.config.js]: Will you expand upon why this configuration is required? Was it to address a specific error? 

Is there anything blocking us from utilized auto-linking? 

---

Tangentially related: I'm not entirely sure we need the existing configuration in this file. I believe the `project` option is unnecessary as we use a standard `ios` directory location. I believe the pre-existing `dependencies` configuration may actually be Metro bundler [configuration](https://metrobundler.dev/docs/configuration/#extranodemodules). I'll research those two separately.

**Review by dcalhoun — APPROVED** [review]
Looks good! Thank you for completing this. 🎉 

I left a comment to help me better understand the RN CLI configuration changes, but I do not believe it should block merging this. 

I will note that I have not yet tested this on a device, I figured we would begin that after all changes are within a singular branch. 

Lastly, I recommend not squashing and merging this into the feature ongoing feature branch. The commit history could prove useful up until our merge into trunk. I'll leave it for you to decide, though.

File: packages/react-native-editor/react-native.config.js
Comment (geriux) [reply packages/react-native-editor/react-native.config.js]: > Will you expand upon why this configuration is required? Was it to address a specific error?

Of course! It was to address an error when adding the following import:

```
apply from: file("../../../../node_modules/@react-native-community/cli-platform-android/native_modules.gradle"); applyNativeModulesAppBuildGradle(project)
``` 

It hasn't been added in the last few upgrades I guess because of errors appearing due to the auto-linking feature. I remember I saw these errors once when I tried to include this. To align our project with the current React Native files structure, this import was needed as it includes the current [PackageList class](https://github.com/react-native-community/cli/blob/2602f83860f7f834b38d1a51bcafbae3e89bb40b/packages/cli-platform-android/native_modules.gradle#L22) definition. Used in the latest upgrades as well but we couldn't adopt it due to not applying this configuration.

As you can see now in the Kotlin file, [it uses this](https://github.com/WordPress/gutenberg/pull/58551/files#diff-2bba332a9cc79ca4865108adb419241e176da1bcd16bb8f1bb5e4312f43be4caR59) whereas before [we were using](https://github.com/WordPress/gutenberg/pull/58551/files#diff-7eaacb31e4eb9acc12388550403d16bdd0e7bfa60ba442adc0c14ed9abff4c38L325) the `List<ReactPackage>` approach.

When `applyNativeModulesAppBuildGradle` is executed it goes through the project's React Native dependencies to auto-link it, but Gutenberg Mobile doesn't use the auto-linking functionality and uses the manual method instead. For Android, using the built binaries from the React Native Libraries publisher repo ([demo app](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/android/app/build.gradle#L150-L160), [bridge](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-bridge/android/react-native-bridge/build.gradle#L98-L108)) and [manually adding the packages](https://github.com/WordPress/gutenberg/pull/58551/files#diff-2bba332a9cc79ca4865108adb419241e176da1bcd16bb8f1bb5e4312f43be4caR59-R75).

Since I saw `native_modules.gradle` also contains configuration for the new architecture I think it's best to adapt it with this upgrade.

The React Native configuration changes overall are just to disable auto-linking for both platforms (we do it manually for iOS as well).

>I believe the project option is unnecessary as we use a standard ios directory location. I believe the pre-existing dependencies configuration may actually be Metro bundler configuration. I'll research those two separately.

I don't think I fully understand this part, which project option are you mentioning?

**Comment by geriux** [comment]
> Lastly, I recommend not squashing and merging this into the feature ongoing feature branch. The commit history could prove useful up until our merge into trunk. I'll leave it for you to decide, though.

Good idea, how do usually do it? I only see two options available, "Squash and merge" and "Rebase and merge" I'm guessing the latter, is that correct?

Thanks for the review!

**Comment by dcalhoun** [comment]
> Good idea, how do usually do it? I only see two options available, "Squash and merge" and "Rebase and merge" I'm guessing the latter, is that correct?

Correct, the "Rebase and merge" option would preserve the commit history. 👍🏻

File: packages/react-native-editor/react-native.config.js
Comment (dcalhoun) [reply packages/react-native-editor/react-native.config.js]: Thank you for elaborating, it helped me understand the context. 

> The React Native configuration changes overall are just to disable auto-linking for both platforms (we do it manually for iOS as well).

Would you be willing to link to the relevant code for iOS? 

> Is there anything blocking us from utilized auto-linking?

I suppose the answer my above question is that the project's desire/need to avoid a Node.js dependency in the host apps is the blocker for us utilizing auto-linking. I.e., downloading manually configured dependencies from `react-native-libraries-publisher` is the chosen alternative. Is that accurate? 

```suggestion
/**
 * The `null` value for each `platform` below disables auto-linking, as
 * we manually link these library binaries to avoid a Node.js dependency 
 * in the host WordPress app. 
 */ 
```

I suggest expanding the inline comment to clarify which code/lines it references and provide the reasoning. WDYT? Note: I wrote the above suggestion in the GitHub comment box, so there may be incorrect spacing or line lengths that could cause lint errors. 

> I don't think I fully understand this part, which project option are you mentioning?

I was referencing the [`project` configuration](https://github.com/WordPress/gutenberg/blob/5fd4953ada078e923caa99f79553ca7412fec6e4/packages/react-native-editor/react-native.config.js#L15-L18), which may not be necessary as we use the standard `./ios` location for React Native projects. 

---

An aside: Links to PR diffs fail to reveal the highlighted portion if the diff is collapsed from being marked a "Viewed," which I have done. The URL format is also cryptic, providing little discernible information as to the targeted subject (`https://github.com/WordPress/gutenberg/pull/58551/files#diff-7eaacb31e4eb9acc12388550403d16bdd0e7bfa60ba442adc0c14ed9abff4c38L325`). So, I did my best to piece together the reference links from context. I generally avoid linking to portions of a GitHub PR diff because of this shortcoming in its implementation and instead link to [the files](https://github.com/WordPress/gutenberg/blob/8763136dfd2163be4efb76c6ccd40e3aa60e1726/packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.kt#L59-L75) themselves or [individual commits](https://github.com/WordPress/gutenberg/commit/2e2ec267abaeafeb732dd8721d87dbdaf966abb1#diff-2bba332a9cc79ca4865108adb419241e176da1bcd16bb8f1bb5e4312f43be4caR59-R75). It's an annoying limitation.

Comment (geriux) [reply packages/react-native-editor/react-native.config.js]: > Would you be willing to link to the relevant code for iOS?

After checking this, we do have auto-linking for the demo app, I got confused with the configuration we have for Gutenberg Mobile and the XCFramework 😅 . I'll remove the `ios` attributes so the auto-linking keeps working for the demo app.

> I suppose the answer my above question is that the project's desire/need to avoid a Node.js dependency in the host apps is the blocker for us utilizing auto-linking. I.e., downloading manually configured dependencies from react-native-libraries-publisher is the chosen alternative. Is that accurate?

I'm not sure if this is the specific case but I do know in the past the host apps were building a lot of things of Gutenberg Mobile from source like pods and maybe node_modules so it was slowing down other teams that weren't specifically working on Gutenberg.

Maybe we could investigate to enable auto-linking for the demo app but ideally it'd be good to use the same binaries the React Native Libraries publisher has for both cases.

> I suggest expanding the inline comment to clarify which code/lines it references and provide the reasoning. WDYT? Note: I wrote the above suggestion in the GitHub comment box, so there may be incorrect spacing or line lengths that could cause lint errors.

👍 

>I was referencing the [project configuration](https://github.com/WordPress/gutenberg/blob/5fd4953ada078e923caa99f79553ca7412fec6e4/packages/react-native-editor/react-native.config.js#L15-L18), which may not be necessary as we use the standard ./ios location for React Native projects.

That was introduced in https://github.com/WordPress/gutenberg/pull/54077 maybe it'd be worth revisiting if it is still needed or not

Comment (dcalhoun) [reply packages/react-native-editor/react-native.config.js]: > Maybe we could investigate to enable auto-linking for the demo app but ideally it'd be good to use the same binaries the React Native Libraries publisher has for both cases.

Yeah, it makes sense to rely upon a consistent approach. 

> That was introduced in https://github.com/WordPress/gutenberg/pull/54077 maybe it'd be worth revisiting if it is still needed or not

Ah, OK. Thank you for sharing.

**Comment by geriux** [comment]
All CI checks in [Gutenberg Mobile passed](https://buildkite.com/automattic/gutenberg-mobile/builds/8535#_) including the E2E tests so I'll proceed to merge this 🚀

**Pull request merged**
